### DISCUSSION TITLE
Allow admins to delete unwanted/unused tags

### DISCUSSION THREAD

**Discussion by ellativity** [forem/forem#16822] [Feature Requests]
[We've discussed this before](https://github.com/forem/forem/discussions/16261), but we're seeing increases in requests for admins to be able to delete empty tags that they no longer want to appear on the tag page.

I believe this can be [resolved on the member-side with reordering of the tag pag](https://github.com/forem/forem/discussions/16263)e, but it would still be nice to be able to clean up the `admin/content_manager/tags` view from time to time.

**Comment by andypiper** [comment]
I guess a related idea would be an ability for mods to flag or submit tags for consideration for removal.

**Reply by ellativity** [reply]
Yeah I'd love to give mods more options when it comes to communicating about tags in general. What do you think this could look/behave like?

Maybe we could add an option to the mod panel to allow mods to flag a tag as well as adding and removing tags?

**Comment by karvounis** [comment]
Is it possible to also support this functionality through the API? It would make tag management way easier, automated and faster. Cheers

**Reply by ellativity** [reply]
I don't agree that admin actions should be performed by API for convenience. I think that any product update we ship for tag management would be easy and fast to use from within the UI.

Bulk actions like this can conceivably be made via SSH if an admin really wants to make these remote decisions about their community.

As a general rule of thumb, I'd recommend actions like tag management should be done internally whilst an admin is present and able to assess the qualitative impact of their actions. There's a reasonable limit to automation when you're managing a platform containing people's content and interactions.

**Reply by karvounis** [reply]
I totally agree that tag management, and generally actions that admins can perform, are first class citizens and should be treated as such. Permissions to be able to do such actions must be strict and the `Authention/Authorization` layer must be well planned and implemented.

However, I do not believe SSH is more secure when it comes to maintaining modern software and does not scale well. The admin might not have access to the underlying hardware that `Forem` runs on or might not be the same person that is responsible for Operations.

I personally think that it would be a good idea to start treating the `Browser` part of Forem as another interface between the `User` and the  `API`. That way, every action that a user/admin can do through the `Browser`, they can also do through the `CLI` (another interface).

```mermaid
graph LR
    User -->|uses| Browser
    User -->|uses| CLI
    Browser --> Auth[Authentication/Authorization]
    CLI --> Auth
    Auth --> B1[API_replica_1]
    Auth --> B2[API_replica_2]
    Auth --> BX[API_replica_X]
    B1 --> DB[(Database)]
    B2 --> DB
    BX --> DB
```

The `Authentication/Authorization` layer is going to be responsible for approving or rejecting requests and then propagating the request to an `API` replica. The `API` is he only component that is able to talk to the Database, process the request and return a proper response. Using that design, we can scale the API independent of the `Browser` instances and provide the ability for users/admins to automate part of their work in a safe manner.

I would love to hear your thoughts on this!
